cpp/src/rolling/lead_lag_nested_detail.cuh

```diff
@@ -0,0 +1,225 @@
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/detail/scatter.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <vector>
+
+namespace cudf::detail {
+namespace {
+/**
+ * @brief Functor to calculate the gather map used for calculating LEAD/LAG.
+ *
+ * @tparam op Aggregation Kind (LEAD vs LAG)
+ * @tparam PrecedingIterator Iterator to retrieve preceding window bounds
+ * @tparam FollowingIterator Iterator to retrieve following window bounds
+ */
+template <aggregation::Kind op, typename PrecedingIterator, typename FollowingIterator>
+struct lead_lag_gather_map_builder {
+  lead_lag_gather_map_builder(size_type input_size,
+                              size_type row_offset,
+                              PrecedingIterator preceding,
+                              FollowingIterator following)
+    : _input_size{input_size},
+      _null_index{input_size + 1},  // Out of input range. Gather returns null.
+      _row_offset{row_offset},
+      _preceding{preceding},
+      _following{following}
+  {
+  }
+
+  template <aggregation::Kind o = op, CUDF_ENABLE_IF(o == aggregation::LEAD)>
+  size_type __device__ operator()(size_type i)
+  {
+    // Note: grouped_*rolling_window() trims preceding/following to
+    // the beginning/end of the group. `rolling_window()` does not.
+    // Must trim _following[i] so as not to go past the column end.
+    auto following = min(_following[i], _input_size - i - 1);
+    return (_row_offset > following) ? _null_index : (i + _row_offset);
+  }
+
+  template <aggregation::Kind o = op, CUDF_ENABLE_IF(o == aggregation::LAG)>
+  size_type __device__ operator()(size_type i)
+  {
+    // Note: grouped_*rolling_window() trims preceding/following to
+    // the beginning/end of the group. `rolling_window()` does not.
+    // Must trim _preceding[i] so as not to go past the column start.
+    auto preceding = min(_preceding[i], i + 1);
+    return (_row_offset > (preceding - 1)) ? _null_index : (i - _row_offset);
+  }
+
+ private:
+  size_type const _input_size;   // Number of rows in input to LEAD/LAG.
+  size_type const _null_index;   // Index value to use to output NULL for LEAD/LAG calculation.
+  size_type const _row_offset;   // LEAD/LAG offset. E.g. For LEAD(2), _row_offset == 2.
+  PrecedingIterator _preceding;  // Iterator to retrieve preceding window offset.
+  FollowingIterator _following;  // Iterator to retrieve following window offset.
+};
+
+/**
+ * @brief Predicate to find indices at which LEAD/LAG evaluated to null.
+ */
+template <typename GatherMapIter>
+struct is_null_index_predicate_impl {
+  is_null_index_predicate_impl(size_type input_size, GatherMapIter gather_)
+    : null_index{input_size + 1}, gather{gather_}
+  {
+  }
+
+  bool __device__ operator()(size_type i) const { return gather[i] == null_index; }
+
+ private:
+  size_type const null_index;  // Index value to use to output NULL for LEAD/LAG calculation.
```

Comment:
There is some inconsistency here. The class above has member variables with underscore `_` prefix.